phpfpm/phpfpm.go

```diff
@@ -163,36 +165,63 @@ func (p *Pool) Update() (err error) {
 
 	defer fcgi.Close()
 
-	env := map[string]string{
-		"SCRIPT_FILENAME": path,
-		"SCRIPT_NAME":     path,
-		"SERVER_SOFTWARE": "go / php-fpm_exporter",
-		"REMOTE_ADDR":     "127.0.0.1",
-		"QUERY_STRING":    "json&full",
-	}
+	// Process might hang on execution indefinitely there, and block other requests.
+	// This still allows for goroutine to live forever, if either of the requests will not finish.
+	done := make(chan *http.Response, 1)
+	errCh := make(chan error, 1)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	go func(ctx context.Context, ch chan *http.Response, errCh chan error) {
+		env := map[string]string{
+			"SCRIPT_FILENAME": path,
+			"SCRIPT_NAME":     path,
+			"SERVER_SOFTWARE": "go / php-fpm_exporter",
+			"REMOTE_ADDR":     "127.0.0.1",
+			"QUERY_STRING":    "json&full",
+		}
 
-	resp, err := fcgi.Get(env)
-	if err != nil {
-		return p.error(err)
-	}
+		res, err := fcgi.Get(env)
+		if err != nil {
+			errCh <- p.error(err)
+			return
+		}
 
-	defer resp.Body.Close()
+		select {
+		case <-ctx.Done():
+			// request is already done, let's just release response resources
+			res.Body.Close()
+		default:
+			//
+			ch <- res
+		}
 
-	content, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return p.error(err)
-	}
+	}(ctx, done, errCh)
 
-	content = JSONResponseFixer(content)
+	select {
+	case err = <-errCh:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	case resp := <-done:
+		defer resp.Body.Close()
+
+		content, err := io.ReadAll(resp.Body)
```

Comment:
_nitpicking_

we could use:
```
		err = json.NewDecoder(resp.Body).Decode(&p)
```
then we don't need extra buffer to hold contents of the response and remove `res.Body.Close()` altogether. We could then avoid having nested `select` and use only one:
```
select {
    err = <- errCh:
        return err
   case resp := <-done:
      // handle response
  case <-time.After(time.Second*10):
      // handle timeout
}
```

Reply:
Sorry. My mistake. I was under the impression that `NewDecoder()` uses [io.ReadCloser](https://golang.org/pkg/io/#ReadCloser) but it doesn't so ignore my comment.